### PR TITLE
Add C# version (.NET Core 3.1)

### PR DESCRIPTION
**Pull request by cadichris**
Hi there,

The C# conversion was pretty smooth. Below I highlight the key points.

### Program output 

image

### Architecture
- It's a .NET Core v3.1 project
- The testing framework is [xUnit 2.4.0](https://xunit.net/)
- Code is in its own folder
https://github.com/murex/mikado-testbuilders-kata/pull/10#issuecomment-707850857
- `Equals` and `GetHashCode` methods are generated by ReSharper
image

### Transformations
- All the getters `getX()` methods have been transformed into properties `X { get; }`
- `IdGenerator` class is no longer thread-safe, as discussed with @aatwi this is OK. Ahmad was just having fun with the threading aspect 😄 
- Interfaces are prefixed with `I` like in `IBook`
- [_Expression body_](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/expression-bodied-members) syntax is used whenever possible
- `enum` have different CaSe convention, they are Camel Case
- `JsonRepository` is the most _transformed_ code : it uses JSON deserialization based on inner classes
- Some loops are converted to LinQ
  - For instance the loop mentioned in the _Tips to Fix the Bugs_ now uses LinQ 
   image
  - I _think_ that's OK since C# developers are at ease with LinQ
 
### Error handling
When JSON repository cannot access JSON data, it shows an _empty_ report prefixed by an error message. 
I thought it was better than a stack trace.  

image

### Open points
- [ ] In the _Test Data Builders_ markdown doc file, I've read this
image
  - From my researches, Rules are not available on xUnit so I think I can't provide C#-related docs here
- [x] ~From https://github.com/murex/contribution I've read~
  > _Before making any contribution to the Project, please complete and sign the CLA on the last page then scan and send a signed copy of the CLA to: OpenSourceSoftware@murex.com_
  - ~You confirm I need to do that?~
- [x] ~Update the *sneak peek* of the solutions to add C#~
- [x] ~How is this PR linked to #8 ?~
  - ~Does it require rework to accommodate the changes brought by this PR?~

### CODE REVIEW

**Comment by cadichris** [comment]
Hey @aatwi 👋 

I think it's easier to do it on my side.

You mean transitionning from 

image

to 

image

Is that right ?
`csharp/` dir would be under `src/` or at root level as a sibling of `docs/` ?

Also, where will the `.json` file be ?

Have a nice day!

**Comment by cadichris** [comment]
Hi @aatwi 

I've put the csharp version in its own folder

image


I've put the `reposisory.json` file in the project.
So for now it's duplicated...if you want it to be somewhere else, let me know

image

I'll remove the "draft" tag of this PR.

**Comment by aatwi** [comment]
Hi @cadichris , 

Sorry I was late to reply :( 
I was wondering if we put that folder at the root, so we ditch the src folder. 

@philou  WDYT?

**Comment by cadichris** [comment]
Also, so that I don't forget: this PR does not update the *sneak peek* of the solutions in the docs.
So preview of solutions are not available in C#. 

I think I should update it...what do you think?
Or maybe the way the doc is organized will change when we migrate to one folder per language?

I'll add a bullet point in the open point list at the top of this issue.

**Comment by aatwi** [comment]
I think we need to update the MD files under the docs folder at some point. 
Since we are using the docs folder to generate the GitHub pages, we wont be moving those pages to language-based folders.

**Comment by aatwi** [comment]
@cadichris, in reference to the folder structure, the image below shows the structure I have now. 
I created a pull-request and waiting a review from Philippe to merge it into the master  

image

**Comment by philou** [comment]
Hi Christophe, thanks a lot for your submission!

Here are answers to your questions:

> How is this PR linked to #8 ?
> Does it require rework to accommodate the changes brought by this PR?

PR #8 is large and we did not yet find the time to go through it in details. Let's go with the code as it is today.

> In the Test Data Builders markdown doc file, I've read this
> From my researches, Rules are not available on xUnit so I think I can't provide C#-related docs here

The idea is to make sure that the mock is 'uninstalled' after each test. There are idiomatic ways to do this in almost all languages. C++ uses the fact that the destructor of objects allocated on the stack are automatically called. JUnit4 has Rules. I know C# has the Disposable interface, when I used to program in C#, I remember having a base test class containing a list of IDisposable objects to dispose after the execution of each test to make sure objects where correctly cleaned up. There might be new ways to do that. Otherwise, just calling it from tearDown does the job as well.

> From https://github.com/murex/contribution I've read
> Before making any contribution to the Project, please complete and sign the CLA on the last page then scan and send a signed copy of the CLA to: OpenSourceSoftware@murex.com

Yes, you'll have to sign the Murex CLA

**Comment by cadichris** [comment]
I've signed the Murex CLA signature.
I've added the C# sneak peeks in the documentation.

I see there are conflicts in a .sln file. I don't know how to look into that... it looks like it's about a sln file at the root of the repo...but this one might be obsolete now we've migrated to per-language folders.
Let me know if I can help.

> > In the Test Data Builders markdown doc file, I've read this
> > From my researches, Rules are not available on xUnit so I think I can't provide C#-related docs here
> 
> The idea is to make sure that the mock is 'uninstalled' after each test. There are idiomatic ways to do this in almost all languages. C++ uses the fact that the destructor of objects allocated on the stack are automatically called. JUnit4 has Rules. I know C# has the Disposable interface, when I used to program in C#, I remember having a base test class containing a list of IDisposable objects to dispose after the execution of each test to make sure objects where correctly cleaned up. There might be new ways to do that. Otherwise, just calling it from tearDown does the job as well.
 
@philou OK thanks for the explanation. Do you think I need to add something to the docs regarding this topic ?

To summarize, the remaining points seem to be : 
- the conflict on .sln file
- the question about adding docs around the "before / after each" concept

**Comment by aatwi** [comment]
Hello @cadichris , 

We did some changes to the master. 
We created a folder for CPP and another for Java at the root. 
Therefore, we moved the sln file from the root to the subfolder cpp. 

did you do a refresh of the remote branch? 
If you prefer we can schedule a quick call for this :) 

Thanks, 
Ahmad

**Comment by aatwi** [comment]
@cadichris, thanks for the update! 
the code looks good, I will merge it into the master branch 

As for the total, you are right. Seems like we did a round-up somehow.

I will need to check why we did that.
